### PR TITLE
Issue15644ProperMergingOfJava2Permissions

### PR DESCRIPTION
**Pull request by emilytee**

### CODE REVIEW

**Comment by emilytee** [comment]
Fixes an issue in https://github.com/OpenLiberty/open-liberty/issues/13908

**Comment by emilytee** [comment]
#build

File: dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/java2sec/GrantEntry.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+ *
+ */
+public class GrantEntry {
+    
+    String codeBase;
+    String signedBy;
+    private List permissionEntries = new ArrayList();
+
+    
+    final static String NEW_LINE = System.getProperty("line.separator");
```
Comment (utle) [inline dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/java2sec/GrantEntry.java:27]: Looks like we have the following variables define in several classes, can we have a constants class or re-use the same one?

    final static String NEW_LINE = System.getProperty("line.separator");
    final static String QUOTED_STRING = "quoted string";
    final static String PERMISSION_TYPE = "permission type";
    final static String GRANT_KEYWORD = "grant";
    final static String KEYSTORE_KEYWORD = "keystore";
    final static String CODEBASE_KEYWORD = "codeBase";
    final static String PERMISSION_KEYWORD = "permission";
    final static String SIGNEDBY_KEYWORD = "signedBy";
    final static String FILTER_KEYWORD = "filterMask";

File: dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/ClassLoadingServiceImpl.java [hunk 2]
```diff
@@ -728,6 +739,15 @@ public boolean isThreadContextClassLoader(ClassLoader cl) {
...
     public void setSharedLibraryProtectionDomains(Map<String, ProtectionDomain> protectionDomainMap) {
+        
+        if (tc.isDebugEnabled()) {
+            Tr.debug(tc, "protectionDomainMap.size = " + protectionDomainMap.size());
+            for (Map.Entry<String, ProtectionDomain> entry : protectionDomainMap.entrySet()) {
+                Tr.debug(tc, "Key (codeBase)= " + entry.getKey() + ", Value (protectionDomain) = " + entry.getValue());
+            }
+
+        }
+
```
Comment (utle) [inline dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/ClassLoadingServiceImpl.java:750]: Can we have method for this debug so I can be used by other classes?

**Review by utle — COMMENTED** [review]
Can we have some prologs comments to help other folks to understand code, please!!

**Comment by emilytee** [comment]
#15385

**Review by acdemyers — APPROVED** [review]

**Comment by emilytee** [comment]
See APAR processing to be sure it was followed.  https://github.ibm.com/WASL3/WASL3SEC/blob/master/twas-codereview/APARPRocess.md

## APAR review 
Review Date:  1/27/2021

Who attended/reviewed:
- [x ] Alaine
- [ ] Dave
- [x ] Hiroko
- [ ] Teddy 
- [ ] Toshi 
- [x ] Ut
- [ ] Will
- [ ] Zech
#### Operation System:
- [ ] Windows
- [ ] Linux/Unix/AIX
- [ ] zOS
- [x ] all platforms
#### Additional Information:

## Problem Information


### Problem Statement
Java 2 permissions are not getting merged properly.   Specifically the java.policy permissions do not always get merged in with the static permissions from server.xml and permissions.xml

### Problem Cause
The merging in of java.policy was done in the wrong class under the assumption that that class would always be traversed.  The  merging now is done in the PermissionManager which will always be invoked by every IBM classloader.'
## Which FVT buckets will be run:
com.ibm.ws.security.java2sec_fat, as well as all buckets that run with java 2 security enabled.

### Testcase Information
- [ ] creating unit test? no
--  Is this change unit testable (see [ Is my code unit testable? > http://washome.austin.ibm.com/xwiki/bin/view/Main/IsMyCodeUnitTestable] for guidance)?
- [ ] Does it have a unit test?  no
- [ ] creating functional test?  no, already have them
- [ ] Does it have a functional test?  yes
- [ ] 
##  Problem Determination Information

##  Code Review Issues and Resolutions
- [ ] Come back for another review? no

## APAR Closing Information:
What versions are you fixing?   Liberty 21.0.0.2
- [ ] WAS855 
- [ ] WAS90
- [ ] Documentation changes? (yes or no):
- [ ] Is this change enabled with a property? (If yes, write property below)
Targeted fixpack / PTF:  21.0.0.2

- [ ] Security / Integrity APAR?:  no
- [ ] If Sec/Int, code rolled back to previous versions?:   no

## Additional Information.

- [ ] Does this change introduce a custom property? If so, please provide details on name, values (with default) and where set.   no
- [ ] All variables used in the fix are fully initialized?  yes
- [ ] All object references used are first checked for null?  yes
- [ ] The fix only includes changes essential to the fix and if needed, additional trace? Verify that if you have entry trace in a method, that there is exit   yes
trace for that method.
- [ ] Did you do a "diff" of the change to the previous version in CMVC to confirm     no, these files are in github and the latest were used
that no other changes are in the file, no previous fixes removed inadvertently?
- [ ] Did you use the proper "checkout/checkin" procedure and NOT use extract or lock in any fashion?  yes
- [ ] Migration issues: Has any code changed an interface not described in the design or has the behavior of any public method changed?   no
- [ ] Consider performance issues. Are the performance programming guidelines being followed?  n/a
A8:
- [ ] Is Message emitted that includes what went wrong with relevant clues why it went wrong, and a possible user action?   no
- [ ]  If there are platform specific elements to this fix, has the fix been tested on these platforms? no platform specific elements to fix
- [ ]  Are there any hardcoded values, such as dates or certificate expirations?   no
- [ ]  Deprecated apis: does the code utilize any deprecated APIs? If so, is there
a plan to remove the use of these APIs in the next release? Are there issues in
different fixpack levels based on deprecated APIs?  no
- [ ]  Does every file added/modified contain the proper copyright statement?   yes
Github copyright changes look like this....
- [ ]  If the APAR is a backport from a later release, please note the defect which was
backported. In addition, is the original developer of this defect aware of the
backport and agreed there are no further changes required?  n/a
- [ ]  Does this change affect the persistence of the user identity via the LtpaToken(2) cookie or others? If yes, has Thread-Safe issues been considered?  no
- [ ]  Does this APAR apply to Liberty profile?   yes
- [ ]  Does the fixing code in anyway change the realm name? If so, you must consider how this will defined authorizations in the cell and cross cell trust configurations.   no
- [ ]  Multi-threaded issues: Does this code utilize any implementations that are not thread-safe such as HashMap? (If so, change to ConcurrentHashMap)   no
- [ ]  Documentation: For tWAS, be sure to update the class prolog with your APAR comments. For Liberty and tWAS, be sure to use comments liberally. It is best to have verbose comments explaining what was changed in code blocks and decision statements.   yes
- [ ]  Did you update the http://washome.pok.ibm.com/xwiki/bin/view/WASL3PDTeam/securityApars page with your APAR info.   n/a - wiki is gone
- [ ]  Are all product name references in the APAR one of the following?  yes - Liberty
- WebSphere Application Server traditional
- WebSphere Application Server Liberty

**Review by una-tapa — APPROVED** [review]
Thanks, Emily

**Comment by emilytee** [comment]
#build

**Comment by emilytee** [comment]
#build

**Comment by emilytee** [comment]
#build

**Comment by emilytee** [comment]
#build

**Comment by emilytee** [comment]
#build

File: dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/java2sec/GrantEntry.java (hunk 1 above)
Comment (emilytee) [reply dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/java2sec/GrantEntry.java:27]: @utle, a Constants.java is created now to define all the constants above.    Classes referring to those constants are updated to use Constants.XYZ vs XYZ where XYZ is defined with the class.

File: dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/ClassLoadingServiceImpl.java (hunk 2 above)
Comment (emilytee) [reply dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/ClassLoadingServiceImpl.java:750]: Comments have been added where needed, including all new classes introduced.

**Comment by emilytee** [comment]
#build

**Comment by emilytee** [comment]
#build

**Comment by emilytee** [comment]
#run-libby-bot

**Pull request merged**